File: lib/digest_renderer.rb
# frozen_string_literal: true

require 'erb'

class DigestRenderer
  TEMPLATE = %(
    Your daily Hacker News digest:
    <br>
    <% for @post in @posts %>
      <p>
        <%= ERB::Escape.html_escape(@post['title']) %>
        <br>
        <%= @post['points'] %> points -
        <% if @post['url'] %>
          <a href="<%= @post['url'] %>">
            link
          </a> -
        <% end %>
        <a href="https://news.ycombinator.com/item?id=<%= @post['objectID'] %>">
          comments
        </a>
      </p>
    <% end %>
    <br>
    To unsubscribe, reply to this email.
  )
  private_constant :TEMPLATE

  def initialize(posts:, date:)
    @date = date
    @posts = posts
  end

  def subject
    "Hacker News Digest for #{@date.getutc.strftime('%b %-d, %Y')}"
  end

  def content
    ERB.new(TEMPLATE, trim_mode: '>-').result(binding)
  end
end

File: rust/src/digest_renderer.rs
use askama::Template;
use chrono::{DateTime, Utc};
use crate::types::Post;
use anyhow::{Result, Context};

#[derive(Template)]
#[template(path = "digest.html")]
struct DigestTemplate<'a> {
    posts: &'a [Post],
}

pub struct DigestRenderer<'a> {
    posts: &'a [Post],
    date: DateTime<Utc>,
}

impl<'a> DigestRenderer<'a> {
    pub fn new(posts: &'a [Post], date: DateTime<Utc>) -> Self {
        Self { posts, date }
    }

    pub fn subject(&self) -> String {
        format!("Hacker News Digest for {}", self.date.format("%b %-d, %Y"))
    }

    pub fn content(&self) -> Result<String> {
        let tmpl = DigestTemplate { posts: self.posts };
        tmpl.render().context("Failed to render digest template")
    }
}
